aten/src/ATen/cuda/CUDABlas.cpp

```diff
@@ -0,0 +1,325 @@
+  }
+  if (error == CUBLAS_STATUS_INVALID_VALUE) {
+    return "CUBLAS_STATUS_INVALID_VALUE";
+  }
+  if (error == CUBLAS_STATUS_ARCH_MISMATCH) {
+    return "CUBLAS_STATUS_ARCH_MISMATCH";
+  }
+  if (error == CUBLAS_STATUS_MAPPING_ERROR) {
+    return "CUBLAS_STATUS_MAPPING_ERROR";
+  }
+  if (error == CUBLAS_STATUS_EXECUTION_FAILED) {
+    return "CUBLAS_STATUS_EXECUTION_FAILED";
+  }
+  if (error == CUBLAS_STATUS_INTERNAL_ERROR) {
+    return "CUBLAS_STATUS_INTERNAL_ERROR";
+  }
+#if CUDA_VERSION >= 6000
+  if (error == CUBLAS_STATUS_NOT_SUPPORTED) {
+    return "CUBLAS_STATUS_NOT_SUPPORTED";
+  }
+#endif
+#if CUDA_VERSION >= 6050
+  if (error == CUBLAS_STATUS_LICENSE_ERROR) {
+    return "CUBLAS_STATUS_LICENSE_ERROR";
+  }
+#endif
+  return "<unknown>";
+}
+
+static cublasOperation_t _cublasOpFromChar(char op) {
+  switch (op) {
+    case 'n':
+    case 'N':
+      return CUBLAS_OP_N;
+    case 't':
+    case 'T':
+      return CUBLAS_OP_T;
+    case 'c':
+    case 'C':
+      return CUBLAS_OP_C;
+  }
+  AT_ERROR(
+      "_cublasOpFromChar input should be 't', 'n' or 'c' but got `", op, "`");
+}
+
+static void _cublasAdjustLdLevel2(int64_t m, int64_t n, int64_t* lda) {
+  // Note: leading dimensions generally are checked that they are > 0
+  // and at least as big the result requires (even if the value won't
+  // be used).
+
+  // Q: Why does Level3 check trans but this doesn't?
+  // A: In level 2, the sizes (m, n) specify the size of A
+  // (independent of trans value). In level 3. the sizes (m, n, k)
+  // specify the sizes of op(A), op(B) where op depend on trans
+  // values.
+  if (n <= 1)
+    *lda = std::max<int64_t>(m, 1);
+}
+
+static void _cublasAdjustLdLevel3(
+    char transa,
+    char transb,
+    int64_t m,
+    int64_t n,
+    int64_t k,
+    int64_t* lda,
+    int64_t* ldb,
+    int64_t* ldc) {
+  bool transa_ = ((transa == 't') || (transa == 'T'));
+  bool transb_ = ((transb == 't') || (transb == 'T'));
+
+  // Note: leading dimensions generally are checked that they are > 0
+  // and at least as big the result requires (even if the value won't
+  // be used).
+  if (n <= 1)
+    *ldc = std::max<int64_t>(m, 1);
+
+  if (transa_) {
+    if (m <= 1)
+      *lda = std::max<int64_t>(k, 1);
+  } else {
+    if (k <= 1)
+      *lda = std::max<int64_t>(m, 1);
+  }
+
+  if (transb_) {
+    if (k <= 1)
+      *ldb = std::max<int64_t>(n, 1);
+  } else {
+    if (n <= 1)
+      *ldb = std::max<int64_t>(k, 1);
+  }
+}
+} // anonymous namespace
+
+namespace at {
+namespace cuda {
+namespace blas {
+
+/* LEVEL 3 BLAS FUNCTIONS */
+
+#define GEMM_CHECK_ARGVALUES(Dtype)           \
+  do {                                        \
+    CUDABLAS_NONNEGINT_CHECK(gemm<Dtype>, m); \
+    CUDABLAS_NONNEGINT_CHECK(gemm<Dtype>, n); \
+    CUDABLAS_NONNEGINT_CHECK(gemm<Dtype>, k); \
+    CUDABLAS_POSINT_CHECK(gemm<Dtype>, lda);  \
+    CUDABLAS_POSINT_CHECK(gemm<Dtype>, ldb);  \
+    CUDABLAS_POSINT_CHECK(gemm<Dtype>, ldc);  \
```

Comment:
Cool, I don't think we were doing zero or one checks before, but I agree that the gemm spec requires them.

Reply:
resolving as nop